### PR TITLE
Add datadog timing for requests to s3 blobdb

### PR DESCRIPTION
**Pull request by dannyroberts**
Opening for review of the general tactic. (There are other things we could do like log to a file and read from the file, or log at the boto library HTTP request level with some monkey patching, but this seemed like this simplest.)

If we're happy with the general implementation, I'll test on staging and then we can roll this out.

### CODE REVIEW

File: corehq/blobs/s3db.py [hunk 1]
```diff
@@ -40,6 +41,24 @@ def __init__(self, config):
         # https://github.com/boto/boto3/issues/259
         self.db.meta.client.meta.events.unregister('before-sign.s3', fix_s3_host)
 
+    def report_timing(self, action):
+        blobdb = self
+        timer = TimingContext()
+        original_stop = timer.stop
+
+        def new_stop(timer, name=None):
```
Comment (millerdev) [inline corehq/blobs/s3db.py]: I'm guessing `timer` here is conventionally called `self`? I think this is not what you want:

```py
class Context(object):
    def stop(self, name=3):
        print(self, name)

timer = Context()
old_stop = timer.stop

def new_stop(self, name=5):
    old_stop(name)
    print(self, name)

timer.stop = new_stop
timer.stop(7)
```
output:
```
(<__main__.Context object at 0x100789910>, 5)
(7, 5)
```

In other words, this line should say `def new_stop(name=None):`.

**Review by millerdev — COMMENTED** [review]
The general approach looks good. I think the `new_stop` signature needs to be fixed.

**Review by snopoke — COMMENTED** [review]
Might be simpler to use the datadog api for this:
```
with statsd.timed(metric_name, tags):
    ....
```

https://github.com/DataDog/datadogpy/blob/master/datadog/dogstatsd/base.py#L210

Or you could use the `TimedContextManagerDecorator` directly to simplify the interface so that you don't have to supply the tags directly

**Comment by millerdev** [comment]
`TimedContextManagerDecorator` uses the `timing` metric, which calculates a histogram.  Datadog's histograms are not very useful to us because they are calculated on a fixed flush interval (10s), which requires a fairly high input frequency to yield meaningful results. Maybe we should use our histogram approximation ([`bucket_value`](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/datadog/utils.py#L89))?

Per Simon's suggestion, it also would be nice to move `report_timing` into `corehq.util.datadog` so it can be used in other places.

**Comment by snopoke** [comment]
Thanks @millerdev, good point. We should upgrade to DD Agent v6 which
reportedly has support for global histograms.

**Comment by dannyroberts** [comment]
Oh, I didn't know about (or forgot about) `statsd.timed`. From https://docs.datadoghq.com/developers/metrics/histograms/, I would think it's exactly what we want. Since it stores both average and count (among other things), I'm assuming it rolls up averages correctly (by averaging w/  sum(average * count) / sum(count), and not simply average of averages / count of averages). I'd think the only thing we lose I think is granularity under 10s, which we aren't concerned about. Do you have an example of it being useless?

**Comment by millerdev** [comment]
@dannyroberts see [Understanding histograms](https://confluence.dimagi.com/pages/viewpage.action?pageId=29559520#TroubleshootingPasteboard/HQchatdumpingground-Datadog) on confluence. There's one article by Datadog and two links to PRs with comments that explain how it relates to us.

**Comment by dannyroberts** [comment]
Yup, you're totally right. I made the same mistake about the histogram `avg` metric as many before me :) Thanks for the links.

I'll update the approach to be something that's more reusable.

**Comment by dannyroberts** [comment]
Oh, and i think checking out other timings e.g. (`nginx.timings`), I should be using a gauge metric, not the counter metric.

File: corehq/blobs/s3db.py (hunk 1 above)
Comment (dannyroberts) [reply corehq/blobs/s3db.py]: Thanks, good catch

**Comment by millerdev** [comment]
> I should be using a gauge metric, not the counter metric.

The guage metric only records one (the last) value per flush interval (10s). For S3 timings I think that will not give a good representation of timings. I'm not sure if you were planning to use that in conjunction with [`bucket_value`](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/datadog/utils.py#L89); if not I think `bucket_value` with a counter is what you want in this PR.

**Comment by dannyroberts** [comment]
Okay, @millerdev I think this is good to go now. In the final version, I went with the bucketing approach and gave up hope to be able to graph the average. This is more meaningful than an average anyway. I also pulled out the portion of the work that could conceivably be useful elsewhere for this.

**Review by millerdev — APPROVED** [review]
Looks great!

**Comment by millerdev** [comment]
Tests failed with 
```
ERROR: corehq.apps.sms.tests.test_backends:AllBackendTest.test_smsgh_inbound_sms
...
CaseNotFound: {'msg': '', 'status_int': None, 'response': None}
-------------------- >> begin captured stdout << ---------------------
<function test_smsgh_inbound_sms at 0x7fe061745050> failed with the following settings:
settings.TESTS_SHOULD_USE_SQL_BACKEND = False
```
Kicked build.

**Pull request merged**